Simbody/src/GeneralForceSubsystem.cpp

```diff
@@ -34,12 +34,244 @@
+                    if (impl.dependsOnlyOnPositions()) {
+                        impl.calcForce(*m_s, *m_rigidBodyForceCache, *m_particleForceCache, *m_mobilityForceCache);
+                    } else { // ordinary velocity dependent force
+                        impl.calcForce(*m_s, *m_rigidBodyForces, *m_particleForces, *m_mobilityForces);
+                    }
+                }
+            } else {
+                // Process a single parallel force. Subtract 1 from threadIndex b/c
+                // we use 0 for the non-parallel forces.
+                const auto& impl =
+                    m_enabledParallelForces->getElt(threadIndex-1)->getImpl();
+                if (impl.dependsOnlyOnPositions()) {
+                    impl.calcForce(*m_s, m_rigidBodyForceCacheLocal.upd(), m_particleForceCacheLocal.upd(), m_mobilityForceCacheLocal.upd());
+                } else { // ordinary velocity dependent force
+                    impl.calcForce(*m_s, m_rigidBodyForcesLocal.upd(), m_particleForcesLocal.upd(), m_mobilityForcesLocal.upd());
+                }
+            }
+            break;
+
+        case NonCached:
+            if (threadIndex == NONPARALLELFORCESINDEX) {
+                // Process all non-parallel forces.
+                for (Force* force : *m_enabledNonParallelForces) {
+                    const auto& impl = force->getImpl();
+                    if (!impl.dependsOnlyOnPositions()) {
+                        impl.calcForce(*m_s,
+                                *m_rigidBodyForces, *m_particleForces,
+                                *m_mobilityForces);
+                    }
+                }
+            } else {
+                // Process a single parallel force. Subtract 1 from index b/c
+                // we use 0 for the non-parallel forces.
+                const auto& impl =
+                    m_enabledParallelForces->getElt(threadIndex-1)->getImpl();
+                    if (!impl.dependsOnlyOnPositions()) {
+                        impl.calcForce(*m_s,
+                                m_rigidBodyForcesLocal.upd(), m_particleForcesLocal.upd(),
+                                m_mobilityForcesLocal.upd());
+                    }
+            }
+            break;
+        }
+    }
+    void finish() override {
+      // using namespace std;
+      // cout <<  m_rigidBodyForcesLocal.get() << endl;
+
+        // Add in this thread's contribution.
+        *m_rigidBodyForces += m_rigidBodyForcesLocal.get();
+        *m_particleForces += m_particleForcesLocal.get();
+        *m_mobilityForces += m_mobilityForcesLocal.get();
+
+        if (m_mode == CachedAndNonCached) {
+            *m_rigidBodyForceCache += m_rigidBodyForceCacheLocal.get();
+            *m_particleForceCache += m_particleForceCacheLocal.get();
+            *m_mobilityForceCache += m_mobilityForceCacheLocal.get();
+        }
+    }
+
+private:
+
+    ParallelExecutor& m_executor;
+
+    Mode m_mode;
+
+    const State* m_s;
```

Comment:
Could be `ReferencePtr<const State> m_state;`. That would self-initialize to `nullptr`.